File: LeetCode/0372. Super Pow/Solution.java
/**
 * Created by Inno Fang on 2018/3/19.
 */

class Solution {

    /**
     * 54 / 54 test cases passed.
     * Status: Accepted
     * Runtime: 6 ms
     * @param a
     * @param b
     * @return
     */
    public int superPow(int a, int[] b) {
        int mod = 1337;
        int result = 1;
        for (int i = b.length - 1; i >= 0; i--) {
            result = result * pow(a, b[i]) % mod;
            a = pow(a, 10);
        }
        return result;
    }

    /**
     * 快速幂
     * @param a
     * @param b
     * @return
     */
    public int pow(int a, int b) {
        int mod = 1337;
        a %= mod;
        int ans = 1;
        while (b != 0) {
            if ((b & 1) == 1) ans = (ans * a) % mod;
            a = (a * a) % mod;
            b >>= 1;
        }
        return ans;
    }
}

File: LeetCode/0372. Super Pow/solution.cpp
/**
 * 55 / 55 test cases passed.
 * Runtime: 12 ms
 * Memory Usage: 11.4 MB 
 */
class Solution {
public:
    const int mod = 1337;
    int superPow(int a, vector<int>& b) {
        if (b.empty()) return 1;
        int last = b.back(); b.pop_back();
        return quick_mod(a, last) % mod * quick_mod(superPow(a, b), 10) % mod;
    }
    size_t quick_mod(size_t x, int y) {
        size_t ret = 1;
        while (y) {
            if (y & 1) ret = (ret * x) % mod;
            x  = x * x % mod;
            y >>= 1;
        }
        return ret;
    }
};

/**
 * 55 / 55 test cases passed.
 * Runtime: 8 ms
 * Memory Usage: 11.4 MB 
 */
class Solution2 {
public:
    const int mod = 1337;
    int superPow(int a, vector<int>& b) {
        int ans = 1;
        for (int i = b.size() - 1; i >= 0; --i) {
            ans = ans * quick_mod(a, b[i]) % mod;
            a = quick_mod(a, 10);
        }
        return ans;
    }
    size_t quick_mod(size_t x, int y) {
        size_t ret = 1;
        while (y) {
            if (y & 1) ret = (ret * x) % mod;
            x  = x * x % mod;
            y >>= 1;
        }
        return ret;
    }
};
